### PR TITLE
Add track initialization to demo loop

### PR DESCRIPTION
**Pull request by amandalund**
This incorporates the initialization of new tracks from primaries and secondaries into the demo loop. Other changes include:
- Fusion of `LDemoInterface` and `TrackInterface`
- Some fixes to the demo loop stepping algorithm
- Setting the lower applicability of models to zero. Since cross section tables might extend beyond the lower applicable energy of the model, it was possible to sample a process for which there was no corresponding model at the given energy. Lowering the model's limit ensures that a model will always be selected for a sampled process.

### CODE REVIEW

**Comment by sethrj** [comment]
Sorry for the delay @amandalund  , I'll be reviewing this today. Looks like there's a minor conflict in the SB model.

**Comment by amandalund** [comment]
thanks @sethrj! I'll fix that up.

File: src/physics/base/PhysicsTrackView.i.hh [hunk 1]
```diff
@@ -341,12 +344,11 @@ PhysicsTrackView::make_model_finder(ParticleProcessId ppid) const
  */
 CELER_FUNCTION real_type PhysicsTrackView::range_to_step(real_type range) const
 {
-    CELER_EXPECT(range > 0);
```
Comment (sethrj) [inline src/physics/base/PhysicsTrackView.i.hh:344]: Should we leave this in with `>=`? The range should never be negative at least?

File: src/physics/base/PhysicsStepUtils.i.hh [hunk 2]
```diff
@@ -222,6 +228,9 @@ select_process_and_model(const ParticleTrackView& particle,
     // Nonzero MFP to interaction -- no interaction model
     CELER_EXPECT(physics.interaction_mfp() <= 0);
 
+    if (physics.macro_xs() == 0)
+        return {};
```
Comment (sethrj) [inline src/physics/base/PhysicsStepUtils.i.hh:232]: There are a lot of new edge cases in this update (I assume because of situations you encountered while actually transporting) but I don't see any new tests in PhysicsStepUtils. Is it possible for you to reproduce these in unit tests? 😬

**Review by sethrj — CHANGES_REQUESTED** [review]
Thanks @amandalund , major props for getting this to work! I think a few more comments about and unit tests for the edge cases would be really helpful going forward: I think once we get the loop working and tested and stable we'll want to explore if there are more sensible ways to structure the sim state/action results/zero values/other exceptions to reduce the number of "if" statements checking for these edge cases.

File: src/physics/base/PhysicsParams.cc [hunk 3]
```diff
@@ -233,7 +232,7 @@ void PhysicsParams::build_ids(const ParticleParams& particles,
         pgroup.models    = model_groups.insert_back(temp_model_groups.begin(),
                                                  temp_model_groups.end());
         // NOTE: data tables will be assigned later
-        CELER_ASSERT(pgroup);
+        CELER_ASSERT(process_to_models.empty() || pgroup);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/base/PhysicsParams.cc:235]: Oh you know what, we might not have talked about this one. This would be ok for particles that don't have any processes defined, i.e. any logged on line 176.

File: src/physics/base/PhysicsStepUtils.i.hh [hunk 4]
```diff
@@ -71,6 +71,12 @@ calc_tabulated_physics_step(const MaterialTrackView& material,
     }
     physics.macro_xs(total_macro_xs);
 
+    real_type interaction_length = inf;
+    if (total_macro_xs > 0)
+    {
+        interaction_length = physics.interaction_mfp() / total_macro_xs;
+    }
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/base/PhysicsStepUtils.i.hh:78]: uhhh you're right 🤦

File: src/random/distributions/UniformRealDistribution.i.hh [hunk 5]
```diff
@@ -31,6 +31,7 @@ UniformRealDistribution<RealType>::UniformRealDistribution(real_type a,
                                                            real_type b)
     : a_(a), delta_(b - a)
 {
+    CELER_EXPECT(a <= b);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/random/distributions/UniformRealDistribution.i.hh:34]: Yeah I actually think I added this back in when I was debugging #245. I don't think `min_energy_fraction` should ever be greater than `max_energy_fraction` now in those energy distributions, should it?

File: app/demo-loop/KernelUtils.i.hh [hunk 6]
```diff
@@ -32,76 +31,133 @@ CELER_FUNCTION void calc_step_limits(const GeoTrackView&      geo,
...
+        model         = ppid_mid.model;
     }
-    return step;
+    phys.model_id(model);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Select the model for the discrete interaction.
+ * Apply secondary cutoffs and process interaction change.
```
[the remark this answers is not in the document]
Comment (amandalund) [reply app/demo-loop/KernelUtils.i.hh:119]: No, it's not guaranteed to be a post-interaction track. If it reached a boundary, it will have `interaction.action == Action::entered_volume` and `action_unchanged(...)` will return true. Looking at this again, I think I need to fix the `Interaction` for boundary-crossing tracks -- as-is it seems like the energy deposition could be nonzero.

File: src/physics/em/detail/BetheHeitler.cu [hunk 7]
```diff
@@ -53,9 +54,10 @@ bethe_heitler_interact_kernel(const BetheHeitlerPointers                bh,
                              particle.particle_id(),
                              material.material_id(),
                              tid);
+    SimTrackView     sim(model.states.sim, tid);
 
     // This interaction only applies if the Bethe-Heitler model was selected
-    if (physics.model_id() != bh.model_id)
+    if (physics.model_id() != bh.model_id || !sim.alive())
```
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/BetheHeitler.cu:60]: This is for when there are more tracks than primaries. While all tracks initially have the sim state initialized with `alive = false`, we don't do the same for the physics state and model ID -- maybe we should instead of adding this check?

For active tracks you're right. If a track escapes the geometry it's killed, and it will have an invalid `ModelId` in that case (as in the case where it crosses a boundary) because the mfp will be nonzero. Stopped particles that can't undergo a discrete interaction are killed, and those will have an invalid `ModelId` since they have zero macro xs. Particles that are absorbed are killed post-interaction, so not relevant here.

File: app/demo-loop/KernelUtils.i.hh [hunk 8]
```diff
@@ -32,76 +31,133 @@ CELER_FUNCTION void calc_step_limits(const GeoTrackView&      geo,
...
+
+    // Calculate energy loss over the step length
+    auto eloss = calc_energy_loss(particle, phys, step);
+    *edep += eloss.value();
+    particle.energy(
+        ParticleTrackView::Energy{particle.energy().value() - eloss.value()});
+
+    // Reduce the remaining mean free path
+    real_type mfp = phys.interaction_mfp() - step * phys.macro_xs();
+    phys.interaction_mfp(soft_zero(mfp) ? 0 : mfp);
```
[the remark this answers is not in the document]
Comment (amandalund) [reply app/demo-loop/KernelUtils.i.hh:104]: Yeah, seems like due to limited precision the mfp calculated here can be slightly less than or greater than zero.

File: src/physics/base/PhysicsStepUtils.i.hh (hunk 2 above)
Comment (amandalund) [reply src/physics/base/PhysicsStepUtils.i.hh:232]: Good point, I'll see what I can put together.

File: src/random/distributions/UniformRealDistribution.i.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/random/distributions/UniformRealDistribution.i.hh:34]: No, you're right... I suspect that on an earlier refactoring it made sense, but now not so much. Let's leave it in to catch any potential unexpected situations, and we can delete it again if we find another good reason.

File: src/physics/em/detail/BetheHeitler.cu (hunk 7 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/detail/BetheHeitler.cu:60]: Good point, or maybe we should put this into the `select_model` bit: instead of `return`ing if `!sim.alive()`, clear the model ID so it's not selected.

File: src/physics/base/PhysicsParams.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/base/PhysicsParams.cc:235]: [for the record] We discussed in Monday's meeting: this is essentially because of an inconsistency between the listed particles and the processes written to the file.

File: app/demo-loop/KernelUtils.i.hh (hunk 6 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply app/demo-loop/KernelUtils.i.hh:119]: Yeah, we can definitely have energy deposition (along-step deposition) when hitting a boundary.

File: src/physics/em/detail/BetheHeitler.cu (hunk 7 above)
[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/BetheHeitler.cu:60]: I think we would need to set `model_id = ModelId{}` for all tracks during initialization since we launch the interaction kernels with the max number of tracks and should exit if the track is empty.

[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/detail/BetheHeitler.cu:60]: Is initialization the right place? Is that going to apply to empty slots as the cascade dies out, in addition to the very first one? I was thinking in the `along_and_post_step_kernel` would work, since that's before the interaction kernels (where the model ID gets set for the active tracks).

[the remark this answers is not in the document]
Comment (amandalund) [reply src/physics/em/detail/BetheHeitler.cu:60]: Setting it at initialization should be alright.. the physics state is cleared at the end of the post-processing kernel if a discrete interaction occurred, and new tracks from primaries/secondaries are initialized with an empty `ModelId`. But you're right, I think clearing it in `along_and_post_step_kernel` would also work and would be easier.

**Comment by amandalund** [comment]
Thanks @sethrj, I made some changes based on our discussion and updated the physics tests. Let me know if you notice anything I missed.

File: src/physics/base/PhysicsStepUtils.i.hh (hunk 2 above)
Comment (amandalund) [reply src/physics/base/PhysicsStepUtils.i.hh]: This particular edge case was a side effect of an error in the discrete interaction sampling -- I was only resetting the mean free path in post-processing if a discrete interaction occurred, but it also needs to be reset if a process was sampled but the discrete interaction was rejected in the integral approach. Fixed that and removed this.

**Review by sethrj — APPROVED** [review]
Excellent work, @amandalund !
